Enqueue Google Fonts instead of @import. Also Set Font Display

AlexGStapleton opened the pull request:
This PR enqueues fonts instead of `@import`ing them.

It also sets Font Display which is required to avoid [a Google performance flag](https://web.dev/font-display/). The value is based on https://github.com/siteorigin/so-widgets-bundle/pull/1072/files

[Related](https://github.com/siteorigin/vantage/pull/381)

AlexGStapleton committed:
Google Fonts: Set Font Display

AlexGStapleton committed:
Prevent @Import of Google Fonts

AlexGStapleton committed:
Enqueue all Google Fonts and add support for variants and subsets

- Supports default fonts.
- The all subset will only be set if the font being loaded is a default font.

Misplon commented:
@AlexGStapleton With Corp and the Site Title font set before switching to this branch I'm seeing `Notice: Array to string conversion in /Users/amisplon/Sites/siteorigin/wp-content/themes/siteorigin-corp/inc/settings/inc/webfont_manager.php on line 31`.

My import is as follows: `<link rel="stylesheet" id="siteorigin-google-web-fonts-css" href="//fonts.googleapis.com/css?family=ABeeZee%3Aregular%7COpen+Sans%3A300%2C400%2C600%7CMontserrat%3A400&amp;subset&amp;display=block&amp;ver=5.7" type="text/css" media="all">`. This means we're getting the same menu font change issue discussed on Slack.

<img width="580" alt="weight" src="https://user-images.githubusercontent.com/789159/112997031-079b1980-916d-11eb-80f7-e3c889adebc0.png">

AlexGStapleton commented:
I'm unable to replicate that error. Can you please clarify your exact font settings?

Misplon commented:
I can initiate the notice without setting any fonts.

<img width="1760" alt="Customize__SiteOrigin_–_Just_another_WordPress_site" src="https://user-images.githubusercontent.com/789159/113005796-0ec62580-9175-11eb-9fd5-ac6aaa2ac063.png">

AlexGStapleton commented:
Unfortunately, I'm still not getting that error with the same settings. Regardless, I'm going to reply with a suggestion. Please try it and let me know (or commit it) if it helps.

AlexGStapleton commented on inc/webfont_manager.php:
```diff
@@ -20,13 +20,15 @@ static function single() {
+				'subset' => $subset,
+			);
+		} else {
+			$this->fonts[ $name ] = array_merge( $this->fonts[$name], $weights );
+			$this->fonts[ $name ] = array_unique( $this->fonts[$name] );
```
```suggestion
			$this->fonts[ $name ] = array_merge( $this->fonts[$name], $weights );

			if ( is_array( $this->fonts[$name] ) ) {
				$this->fonts[$name] = array_unique( $this->fonts[$name] );
			}
```

@Misplon  Try changing the highlighted lines to this. Does that prevent the notice from appearing, and working as intended?

Misplon replied:
Assuming this change is correct then, unfortunately, this hasn't helped.

<img width="1169" alt="settings" src="https://user-images.githubusercontent.com/789159/113280720-f5dc8200-92e4-11eb-8388-28dfdeabdfa3.png">

AlexGStapleton replied:
Yeah, that's correct. :(
I've set up a fresh install and I'm not able to replicate this error on that install either - I'm using Corp there also. Are you able to replicate this on another install?

Misplon commented:
@AlexGStapleton Are we expecting this PR to deal with the Corp, Montserrat issue at this stage? I've uploaded this branch to the demo server. If you set the site title font you'll see we lose the menu font-weight that was there before. My bad if we've already covered this on Slack.

Misplon commented:
I haven’t replicated the notice yet, no but the issue of the menu font
weight being lost in Corp after a font is set is happening locally and on
the demo server.

On Thu, 01 Apr 2021 at 16:18, Alex S ***@***.***> wrote:

> *@AlexGStapleton* commented on this pull request.
> ------------------------------
>
> In inc/webfont_manager.php
> <https://github.com/siteorigin/settings/pull/69#discussion_r605694997>:
>
> > +			$this->fonts[ $name ] = array_merge( $this->fonts[$name], $weights );
> +			$this->fonts[ $name ] = array_unique( $this->fonts[$name] );
>
> Yeah, that's correct. :(
> I've set up a fresh install and I'm not able to replicate this error on
> that install either - I'm using Corp there also. Are you able to replicate
> this on another install?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/siteorigin/settings/pull/69#discussion_r605694997>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGAVJ3M7DCDQGTVKMQXFJDTGR6DPANCNFSM4Z7SVWEQ>
> .
>
-- 
Sent from Gmail Mobile

AlexGStapleton commented:
I've looked into Corp losing the menu weight, and it's unrelated. The menu in Corp has a font-wight of 700 and that font-weight is loaded by the site title so if that font is changed, the 700 font weight version of the font isn't able to be loaded.

https://github.com/siteorigin/siteorigin-corp/blob/develop/inc/settings.php#L542-L572

I think we should add 700 for body font weight (Corp PR). We should also add 700 as a selected font weight for the body font (separate PR for settings library and should ideally be adding a filter and an additional Corp PR). That way the font-weight should be used regardless of the selected font if the font itself supports it. Thoughts?

AlexGStapleton committed:
Improve font variant merging

The font names will always match so we should instead opt to merge the variants directly.

AlexGStapleton commented:
The last commit should prevent the notice from occurring.

Misplon commented:
Was there a commit you meant to push, mate?

Misplon commented:
Sounds good on the Corp suggestions, please, proceed.

AlexGStapleton commented:
> Was there a commit you meant to push, mate?

Whoops! Just pushed it.

AlexGStapleton committed:
Font arrays: Resolve notice

Notice: Array to string conversion in /wp-content/themes/siteorigin-corp/inc/settings/inc/webfont_manager.php on line 36
